#!/usr/bin/env ruby
require 'bundler'
Bundler.setup

require 'octokit'
require 'active_support'
require 'active_support/core_ext/date/calculations'
require 'active_support/core_ext/integer/time'

MESSAGE = <<-BODY
This issue has been automatically closed because of inactivity.

If you can still reproduce this error on the `5-0-stable` branch or on `master`,
please reply with all of the information you have about it in order to keep the issue open.

Thank you for all your contributions.
BODY


def prompt
  print " y/n? "
  STDIN.gets.chomp == "y"
end

def close_issue(client, number)
  client.close_issue('rails/rails', number)
  client.add_comment('rails/rails', number, MESSAGE)
  puts "Closed ##{number}"
end

force = ARGV[0] == '--force'
interactive = ARGV[0] == '-i'

client = Octokit::Client.new(:netrc => true, :netrc_file => '.netrc')

client.auto_paginate = true
issues = client.issues 'rails/rails', state: :open, sort: :updated, direction: :asc
stale_issues = issues.select do |issue|
  issue.labels.detect { |label| label.name == 'stale' }
end

puts "Closing #{stale_issues.size} issues.\n"

stale_issues.each do |issue|
  number = issue.number

  print "Closing issue ##{number}. https://github.com/rails/rails/issues/#{number}"

  if force
    close_issue(client, number)
  elsif interactive && prompt
    close_issue(client, number)
  else
    puts
  end
end
